vumi/middleware/session_length.py

```diff
@@ -0,0 +1,75 @@
+# -*- test-case-name: vumi.middleware.tests.test_session_length -*-
+
+import time
+
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from vumi.message import TransportUserMessage
+from vumi.middleware.base import BaseMiddleware
+from vumi.persist.txredis_manager import TxRedisManager
+
+
+class SessionLengthMiddleware(BaseMiddleware):
+    """ Middleware for storing the session length in the message.
+
+    Session length is stored if the end of the session is reached.
+
+    Configuration option:
+
+    :param dict redis:
+        Redis configuration parameters.
+    :param int timeout:
+        Redis key timeout (in seconds). Defaults to 120.
+    """
+    SESSION_NEW, SESSION_CLOSE = (
+        TransportUserMessage.SESSION_NEW, TransportUserMessage.SESSION_CLOSE)
+
+    @inlineCallbacks
+    def setup_middleware(self):
+        r_config = self.config.get('redis_manager', {})
+        self.redis = yield TxRedisManager.from_config(r_config)
+        self.timeout = self.config.get('timeout', 120)
+
+    @inlineCallbacks
+    def teardown_middleware(self):
+        yield self.redis.close_manager()
+
+    def _set_message_billing_metadata(self, message, field, value):
+        if not message['helper_metadata'].get('billing'):
+            message['helper_metadata']['billing'] = {}
+        message['helper_metadata']['billing'][field] = value
+
+    def _set_session_start_time(self, message, time):
+        self._set_message_billing_metadata(message, 'session_start', time)
+
+    def _set_session_end_time(self, message, time):
+        self._set_message_billing_metadata(message, 'session_end', time)
+
+    @inlineCallbacks
+    def _process_message(self, message, redis_key):
+        if message.get('event_type') == self.SESSION_NEW:
```

Comment:
I think we want to use the `session_event` field. I think `even_type` is a field used for event messages (acks, nacks and delivery reports).